sw/device/lib/crypto/aes_gcm/aes_gcm.c

```diff
@@ -354,3 +373,93 @@ aes_error_t aes_gcm_ghash(const aes_block_t *hash_subkey,
+ */
+static aes_error_t aes_single_block(const aes_params_t params,
+                                    const aes_block_t *input,
+                                    aes_block_t *output) {
+  aes_error_t err = aes_begin(params);
+  if (err != kAesOk) {
+    return err;
+  }
+  err = aes_update(output, input);
+  if (err != kAesOk) {
+    return err;
+  }
+  err = aes_end();
+  if (err != kAesOk) {
+    return err;
+  }
+  return kAesOk;
+}
+
+/**
+ * Implements the GCTR function as specified in SP800-38D, section 6.5.
+ *
+ * The block cipher is fixed to AES. Note that the GCTR function is a modified
+ * version of the AES-CTR mode of encryption.
+ *
+ * @param aes_ctr_params AES parameters for CTR encryption with IV set to the
+ * initial counter block
+ * @param input_len Number of bytes in the input
+ * @param input Pointer to input buffer (may be NULL if input_len is 0)
+ * @param output Pointer to output buffer (same size as input, may be the same
+ * buffer)
+ */
+aes_error_t aes_gcm_gctr(aes_params_t aes_ctr_params, const size_t input_len,
+                         const uint8_t *input, uint8_t *output) {
+  // If the input is empty, the output must be as well. Since the output length
+  // is 0, simply return.
+  if (input_len == 0) {
+    return kAesOk;
+  }
+
+  // NULL pointers are not allowed for nonzero input length.
+  if (input == NULL || output == NULL) {
+    return kAesInternalError;
+  }
+
+  // Check the input parameters.
+  if (aes_ctr_params.mode != kAesCipherModeCtr || !aes_ctr_params.encrypt) {
+    return kAesInternalError;
+  }
+
+  // Calculate the number of blocks needed to represent the input. Since we
+  // checked for empty input above, nblocks must be at least 1.
+  size_t nblocks = get_nblocks(input_len);
+  for (size_t i = 0; i < nblocks; ++i) {
+    // Calculate the number of bytes in the block. All blocks are full except
+    // for (potentially) the last block.
+    size_t nbytes = kAesBlockNumBytes;
+    if ((i == nblocks - 1) && (get_block_offset(input_len) != 0)) {
+      // Last block is not full.
+      nbytes = get_block_offset(input_len);
+    }
+
+    // Retrieve the next block of input. If the block is partial, the input
+    // data will be padded with zeroes.
+    aes_block_t block_in = {.data = {0}};
+    memcpy(block_in.data, &input[i * kAesBlockNumBytes], nbytes);
+
+    // Allocate a buffer for the cipher output.
+    aes_block_t block_out = {.data = {0}};
+
+    // Run the AES-CTR encryption operation on the next block of input.
+    aes_error_t err = aes_single_block(aes_ctr_params, &block_in, &block_out);
+    if (err != kAesOk) {
+      return err;
+    }
+
+    // Copy the result block into the output buffer, truncating some bytes if
+    // the input block was partial.
+    memcpy(&output[i * kAesBlockNumBytes], block_out.data, nbytes);
+
+    // Update the counter block.
+    aes_ctr_params.iv[kAesBlockNumWords - 1] =
```

Comment:
How do we alleviate the very theoretical concern that IV will wrap around after being incremented 2^32 times? Should this function enforce `input_len <= UINT32_MAX`?

Reply:
It's enforced by the top-level encryption function; in fact, the entire byte-length of the input is `<= UINT32_MAX`. We could loosen that bound in the future, but the top-level NIST specification says that implementations must restrict the lengths of plaintext such that the GCTR input is below 2^32 blocks long, so we can expect that length to have been checked by the time we get to GCTR. I'll add an explicit note to the docstring saying we make that assumption, though.